### DISCUSSION TITLE
server + sensors = a system

### DISCUSSION THREAD

**Discussion by jadudm** [18F/imls-pi-stack#47] [Ideas]
This may seem simple, but I don't think we can roll a "v3" without treating the server and sensors as a unit.

## about v2

The currently live version of the sensor platform is completely disconnected from the server. This is because everything is manual:

1. A librarian requests an API key (manual)
2. An admin authorizes the API key to insert data (manual)
3. A librarian sets up a sensor (manual)
4. A librarian enters the API key into the sensor at setup time (manual)

To start scaling, we need to make these things more automated. Possibly all of them.

## server + sensors = a system

The server needs to know about the sensors, and visa-versa. I'm thinking something like this...

1. The SDC sends out a "magic word" for all librarians in their state.
1. The librarian goes to a server (e.g. `sensors.ga.us`).
2. The librarian enters an FCFS Seq Id, a tag, an "owner email address," and the "magic word."  
3. The server provides an installation package for download (~64MB).
    The package is basically an installation image for the sensor, but it includes the metadata (FCFS Seq Id, etc.), so the sensor is "good to go," and can represent itself when submitting data to the server.
4. The librarian sets up the sensor using the package.
5. An admin (in the SDC office) gets notice that a new sensor is live.

The "magic word" is a low-security passphrase to control image creation. Or, it could be a TOTP token. It is, in some form or another, a "shared secret" between the SDCs and the library directors in their state. (I am skeptical of this, but what I'm trying to do is establish a *notional chain of trust* from the SDC to the librarian setting up the sensor unsupervised. I welcome other strategies.)

### fully automated vs manual

The above example is fully automatic. Once we have a server running, librarians can download and set up sensors. The server can provide things like a dashboard ("is my sensor alive?") as well as the ability to download data, etc.

We can also use the server (via API) to set up bulk sensors. It is easy enough to create a small tool that will write the installation package to microSD cards over-and-over. This tool can be driven by a simple CSV file that the SDC creates, driving the setup of (say) 50 devices at a time:

```
ME1234-001,under-desk,yetanothermatt@us.gov
ME1234-002,closet,allthematts@us.gov
ME1234-003,basement,yesamatt@us.gov
...
```
We'll end up needing an admin panel/API for this kind of functionality, but we can do that.

## thoughts?

This idea replaces the following two discussions:

1. [Sensor platform maintenance/OTA updates](https://github.com/18F/imls-pi-stack/discussions/44)
2. [Identifying sensors](https://github.com/18F/imls-pi-stack/discussions/45)

This (short) discussion starter was inspired by a longer, rambling blog over at jadud.com titled [musing on sensor systems](https://jadud.com/p/custompi/).

**Comment by cantsin** [comment]
This is as good as anything we've seen so far. A coupla questions:

1. Do we want to ensure continuity between v2 and v3 versions? I'm kind of inclined to have current pilot devices re-do the installation process so we have a clean(er) slate, but I'm open either way. 
2. "An admin (in the SDC office) gets notice that a new sensor is live." -- how, exactly? Email?

**Reply by jadudm** [reply]
1. At this stage, I'm assuming we are not yet in a world where we have solved in-place updates (fully). This needs its own thread. If the OS stays up-to-date via normal mechanisms, perhaps this next rollout is where we move to containerized apps, so the OS itself has very little "pollution," and our app(s) pack what they need in their images?
2. Yeah. That's a "magic" handwave. In theory, the SDC doesn't *have* to know. It could be they get an email, or it could be the next time they log into the server there's a "messages" space, where they can see system update/alert type notes. TBD?
